### DISCUSSION TITLE
having type problem for create resuable machines with typescript generic

### DISCUSSION THREAD

**Discussion by ali-shabani** [statelyai/xstate#2947] [Q&A]
Hi, I decided to create a reusable machine for my forms. But I have some issues with my typing. Everything looks fine, but I can't figure out why this is not working.

I created my event types as below:
```typescript
export type TouchEvent = { type: "TOUCH" };
export type SubmitEvent<F extends Record<string, any>> = {
  type: "SUBMIT";
} & F;

export type FormEvents<F> = TouchEvent | SubmitEvent<F>;
```

and I created a function to create machine as below:

```typescript
export const createFormMachine = <F>(
  onSubmit: OnSubmit<F>
) =>
  createMachine<PartialFormContext, FormEvents<F>, FormStates>({
    initial: FormState.IDLE,
    states: {
      [FormState.IDLE]: {
        on: {
          TOUCH: FormState.TOUCHED,
        },
      },
  }
})
```

But the typescript compiler complains that there is no `TOUCH` to be used in my state transition.

Below is the exact error that I get from the typescript:
```
Type '{ TOUCH: FormState.TOUCHED; }' is not assignable to type 'TransitionsConfig<MachineContext<FormContext, "">, FormEvents<F>> | undefined'.
  Object literal may only specify known properties, and 'TOUCH' does not exist in type 'TransitionsConfig<MachineContext<FormContext, "">, FormEvents<F>>'.
```

if I remove the type concatenation in `SubmitEvent` and write like this:
```typescript
export type SubmitEvent<F extends Record<string, any>> = {
  type: "SUBMIT";
} & F;

// change above type to below

export type SubmitEvent<F extends Record<string, any>> = {
  type: "SUBMIT";
};
```

Then all errors will perish.
But then I have no way to pass the extra information I need to submit my form when a submit event happens.
below is some screenshot from my VS CODE:
![image](https://user-images.githubusercontent.com/12711392/149815141-cd4f6100-ed1c-4701-a906-fa52f0fc2469.png)

![image](https://user-images.githubusercontent.com/12711392/149814766-96e5224d-0e9f-4e49-ba97-b45fa42619c9.png)

![image](https://user-images.githubusercontent.com/12711392/149814834-4862d512-ae43-450b-baae-d07026ef4839.png)

**Comment by davidkpiano** [comment]
Can you put this into a CodeSandbox?

**Reply by ali-shabani** [reply]
Here you are:
https://codesandbox.io/s/tender-bhaskara-ph673?file=/src/form/form-types.ts
Also, I added you in CodeSandbox and gave you edit access.
